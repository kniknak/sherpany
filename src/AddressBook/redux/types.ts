export enum Status {
    Initial = 1,
    Success,
    Pending,
    Failure,
}

export type Nationality = "CH" | "ES" | "FR" | "GB"

export interface AddressBookState {
    status: Status
    users: FetchAddressBookUser[][]
    page: number
    nationalities: Set<Nationality>
}

export interface FetchAddressBookUser {
    name: {
        first: string
        last: string
    }
    location: {
        street: {
            name: string
            number: number
        }
        city: string
        state: string
        postcode: string
    }
    email: string
    login: {
        username: string
    }
    phone: string
    cell: string
    id: {
        value: string
    }
    picture: {
        thumbnail: string
    }
}

export interface FetchAddressBookPayload {
    users: FetchAddressBookUser[]
    page: number
}
