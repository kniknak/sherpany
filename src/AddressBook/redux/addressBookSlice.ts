import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { AddressBookState, FetchAddressBookPayload, Nationality, Status } from "./types";

export const nationalities: Nationality[] = ["CH", "ES", "FR", "GB"];

export const addressBookInitialState: AddressBookState = {
    status: Status.Initial,
    users: [],
    page: 1,
    nationalities: new Set(nationalities),
};

export const addressBookSlice = createSlice({
    name: "counter",
    initialState: addressBookInitialState,
    reducers: {
        request: state => {
            state.status = Status.Pending;
        },
        success: (state, action: { payload: FetchAddressBookPayload }) => {
            state.status = Status.Success;
            state.users[action.payload.page - 1] = action.payload.users;
        },
        failure: state => {
            state.status = Status.Failure;
        },
        toggleNationality: {
            reducer: (state, action: PayloadAction<Nationality>) => {
                if (state.nationalities.has(action.payload)) {
                    state.nationalities.delete(action.payload);
                } else {
                    state.nationalities.add(action.payload);
                }
            },
            prepare: (nationality: Nationality) => ({ payload: nationality }),
        },
    },
});