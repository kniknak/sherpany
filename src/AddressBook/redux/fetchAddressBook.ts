import { createAction } from "redux-api-middleware";
import { addressBookSlice, nationalities } from "./addressBookSlice";
import { FetchAddressBookPayload, FetchAddressBookUser } from "./types";
import { RootState } from "../../App/store";

export const fetchAddressBook = (page: number) => createAction<Pick<RootState, "addressBook">, Promise<FetchAddressBookPayload>, never>({
    endpoint: state => {
        const nat = [...(state.addressBook.nationalities.size
            ? state.addressBook.nationalities
            : nationalities),
        ]
            .join(",");

        return `https://randomuser.me/api/?seed=sheerpany&page=${page}&results=50&nat=${nat}`;
    },
    method: "GET",
    types: [
        addressBookSlice.actions.request.type,
        {
            type: addressBookSlice.actions.success.type,
            payload: (action, state, response) => (
                response
                    .json()
                    .then<FetchAddressBookPayload>((result: FetchAddressBookResponse) => ({
                        users: result.results,
                        page,
                    }))
            ),
        },
        addressBookSlice.actions.failure.type,
    ],
});

interface FetchAddressBookResponse {
    results: FetchAddressBookUser[]
    info: {
        seed: "sheerpany"
        results: number
        page: number
        version: string
    }
}