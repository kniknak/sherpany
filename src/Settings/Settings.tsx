import React from "react";
import { RootState } from "../App/store";
import { connect, MapStateToProps } from "react-redux";
import Container from "@material-ui/core/Container";
import FormControlLabel from "@material-ui/core/FormControlLabel";
import Checkbox from "@material-ui/core/Checkbox";
import { addressBookSlice, nationalities } from "../AddressBook/redux/addressBookSlice";
import { Nationality } from "../AddressBook/redux/types";

interface StateProps {
    enabledNationalities: Set<Nationality>
}

interface DispatchProps {
    toggleNationality: (nationality: Nationality) => void
}

export type SettingsProps = StateProps & DispatchProps

export class Settings extends React.PureComponent<SettingsProps> {
    private toggleNationality: React.ChangeEventHandler<HTMLInputElement> = event => this.props.toggleNationality(event.target.name as Nationality);

    render() {
        const { enabledNationalities } = this.props;

        return (
            <Container maxWidth="xs">
                {nationalities.map(nationality => (
                    <FormControlLabel
                        key={nationality}
                        control={<Checkbox
                            checked={enabledNationalities.has(nationality)}
                            onChange={this.toggleNationality}
                            name={nationality}
                        />}
                        label={nationality}
                    />
                ))}
            </Container>

        );
    }
}

const mapStateToProps: MapStateToProps<StateProps, {}, Pick<RootState, "addressBook">> = state => ({
    enabledNationalities: state.addressBook.nationalities,
});

const mapDispatchToProps: DispatchProps = {
    toggleNationality: addressBookSlice.actions.toggleNationality,
};

export const SettingsConnected = connect(mapStateToProps, mapDispatchToProps)(Settings);

export default SettingsConnected