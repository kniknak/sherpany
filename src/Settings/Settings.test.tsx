import React from "react";
import { shallow } from "enzyme";
import { Settings, SettingsProps } from "./Settings";
import { Nationality } from "../AddressBook/redux/types";
import FormControlLabel from "@material-ui/core/FormControlLabel";

describe("Settings", () => {
    describe("Settings", () => {
        const defaultProps: SettingsProps = {
            enabledNationalities: new Set<Nationality>(),
            toggleNationality: () => undefined,
        };

        it("renders", () => {
            const subject = shallow(<Settings {...defaultProps}/>);

            expect(subject).toMatchSnapshot();
        });

        it("toggles nationality", () => {
            const props: SettingsProps = {
                ...defaultProps,
                toggleNationality: jest.fn(),
            };

            const subject = shallow(<Settings {...props}/>);

            const event = {
                target: {
                    name: "CF",
                },
            };

            subject.find(FormControlLabel).at(0).prop("control").props.onChange(event);

            expect(props.toggleNationality).toHaveBeenLastCalledWith(event.target.name);
        });
    });
});
