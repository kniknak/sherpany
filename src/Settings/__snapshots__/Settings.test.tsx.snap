// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings Settings renders 1`] = `
<WithStyles(ForwardRef(Container))
  maxWidth="xs"
>
  <WithStyles(ForwardRef(FormControlLabel))
    control={
      <ForwardRef(WithStyles)
        checked={false}
        name="CH"
        onChange={[Function]}
      />
    }
    key="CH"
    label="CH"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    control={
      <ForwardRef(WithStyles)
        checked={false}
        name="ES"
        onChange={[Function]}
      />
    }
    key="ES"
    label="ES"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    control={
      <ForwardRef(WithStyles)
        checked={false}
        name="FR"
        onChange={[Function]}
      />
    }
    key="FR"
    label="FR"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    control={
      <ForwardRef(WithStyles)
        checked={false}
        name="GB"
        onChange={[Function]}
      />
    }
    key="GB"
    label="GB"
  />
</WithStyles(ForwardRef(Container))>
`;
